File: unreliable-udp/receiver.py
#!/usr/bin/env python3

from socket import *
import sys
import signal

if (len(sys.argv) != 3):
    print("Usage: receiver <file name> <timeout (ms)>")
    sys.exit(0)

fileName = sys.argv[1]
timeout = int(sys.argv[2])

addr = ("0.0.0.0", 0)

receiverSocket = socket(AF_INET, SOCK_DGRAM)
receiverSocket.bind(addr)

print('listening on port:', receiverSocket.getsockname()[1])

portfile = open("port", "w")
portfile.write(str(receiverSocket.getsockname()[1])+"\n")
portfile.close()

outputfile = open(fileName, "wb")

numreceived = 0
bytesreceived = 0

data,addr = receiverSocket.recvfrom(16)
packetSize = int(data.decode('utf-8'))

def alarmhandler(signum, frame):
    print(f'{numreceived} {bytesreceived}')
    sys.exit(0)

signal.signal(signal.SIGALRM, alarmhandler)
signal.alarm(int(timeout/1000))

numreceived = 1
bytesreceived = 16

while True:
    data,addr = receiverSocket.recvfrom(packetSize)
    numreceived += 1
    bytesreceived += packetSize
    end = False
    try:
        text = data.decode()
        if (text == ""):
            end = True
    except UnicodeDecodeError:
        end = False
    if (end == True):
        break
    outputfile.write(data)
#end while
outputfile.close()

print(f'{numreceived} {bytesreceived}')

File: unreliable-udp/receiver.cpp
#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/socket.h>
#include <netdb.h>
#include <netinet/in.h>
#include <unistd.h>
#include <iostream>
#include <fstream>


// receiver <file name> <timeout>
int main(int argc, char *argv[]) {
    char filename[100];
    int timeout, sockretval;
    struct sockaddr_in senderaddr;
    unsigned char buf[2048];
    unsigned int recvlen, addrlen; 

    if (argc == 3){
        strcpy(filename, argv[1]);
        timeout = atoi(argv[2]);
    } else {
        printf("Usage: receiver <file name> <timeout> \n");
    }
    
    if ((sockretval = socket(AF_INET, SOCK_DGRAM, 0)) < 0){ 
        perror("socket creation failed"); 
        return 1; 
    }

    // memset((char *)&senderaddr, 0, sizeof(senderaddr));
    senderaddr.sin_family = AF_INET;
    senderaddr.sin_addr.s_addr = htonl(INADDR_ANY);
    senderaddr.sin_port = htons(0);

    if (bind(sockretval, (struct sockaddr *)&senderaddr, sizeof(senderaddr)) < 0) {
        perror("bind failed");
        return 1;
    }

    int portnum = senderaddr.sin_port;

    std::ofstream portfile("port");  
    portfile << portnum << std::endl;  
    portfile.close();


    std::ofstream recvdfile(filename);  
    for (;;) {
        printf("waiting on port %d\n", portnum);
        recvlen = recvfrom(sockretval, buf, 2048, 0, (struct sockaddr *)&senderaddr, &addrlen);
        printf("received %d bytes\n", recvlen);
        if (recvlen > 0){
            buf[recvlen] = 0;
            printf("received message: \"%s\"\n", buf);
        }
        recvdfile << buf;
    }
    recvdfile.close();

    close(sockretval);

    return 0;
}
